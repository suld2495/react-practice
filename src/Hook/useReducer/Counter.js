import React, { useReducer } from 'react'

const reducer = (state, action) => {
    switch(action.type) {
        case 'INCREASE':
            return {
                value: state.value + 1
            }
        case 'DECREASE':
            return {
                value: state.value -1
            }
        default:
            return state;
    }
}

const Counter = () => {
    const [state, dispatch] = useReducer(reducer, { value: 0 });
    const increase = () => {
        dispatch({
            type: 'INCREASE'
        })
    }
    const decrease = () => {
        dispatch({
            type: 'DECREASE'
        })
    }
    return (
        <div>
            {state.value}
            <button onClick={increase}>+1</button>
            <button onClick={decrease}>-1</button>
        </div>
    )
}

export default Counter
