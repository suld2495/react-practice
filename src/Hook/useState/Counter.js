import React, { useState } from 'react'

const Counter = () => {
    // const initialState = someExpensiveComputation();

    function someExpensiveComputation() {
        console.log(1);
        return 0;
    }
 
    // const [value, setValue] = useState(initialState);
    const [value, setValue] = useState(() => {
        return someExpensiveComputation();
    });
    const [name, setName] = useState('');

    return (
        <div>
            <input value={name} onChange={e => setName(e.target.value)}/>
            <p>나의 이름은 {name} 입니다.</p>
            <p>
                현재 카운터 값은 <b>{value}</b>입니다.
            </p>        
            <button onClick={() => setValue(value + 1)}>+1</button>
            <button onClick={() => setValue(value - 1)}>-1</button>
        </div>
    )
}

export default Counter
