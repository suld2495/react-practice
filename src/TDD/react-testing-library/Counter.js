import React, { useCallback, useState } from 'react'

const Counter = () => {
    const [count, setCount] = useState(0);
    const onIncrease = useCallback(
        () => {
            setCount(count + 1);     
        }, 
        [count],
    );
    const onDecrease = useCallback(
        () => {
            setCount(count - 1);
        },
        [count]
    )

    return (
        <div>
            <p>{count}</p>
            <div className="buttons">
                <button onClick={onIncrease}>+</button>
                <button onClick={onDecrease}>-</button>
            </div>
        </div>
    )
}

export default Counter
