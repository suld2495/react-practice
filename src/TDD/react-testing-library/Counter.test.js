import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Counter from './Counter';

describe('<Counter />', () => {
    // test('matches snapshot', () => {
    //     const utils = render(<Counter />);
    //     expect(utils.container).toMatchSnapshot();
    // });

    it('has a number and two buttons', () => {
        render(<Counter />);

        const numberElement = screen.getByText(0);
        expect(numberElement).toBeInTheDocument();

        const plusElement = screen.getByText('+');
        expect(plusElement).toBeInTheDocument();

        const minusElement = screen.getByText('-');
        expect(minusElement).toBeInTheDocument();
    });

    it('increase', () => {
        render(<Counter />);

        const numberElement1 = screen.getByText(0);
        expect(numberElement1).toBeInTheDocument();

        const plusElement = screen.getByText('+');
        expect(plusElement).toBeInTheDocument();

        fireEvent.click(plusElement);

        const numberElement2 = screen.getByText(1);
        expect(numberElement2).toBeInTheDocument();
    });

    it('decrease', () => {
        render(<Counter />);

        const numberElement1 = screen.getByText(0);
        expect(numberElement1).toBeInTheDocument();

        const minusElement = screen.getByText('-');
        expect(minusElement).toBeInTheDocument();

        fireEvent.click(minusElement);

        const numberElement2 = screen.getByText(-1);
        expect(numberElement2).toBeInTheDocument();
    });
});