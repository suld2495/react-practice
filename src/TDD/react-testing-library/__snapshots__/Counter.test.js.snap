// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> matches snapshot 1`] = `
<div>
  <div>
    <p>
      0
    </p>
    <div
      class="buttons"
    >
      <button>
        +
      </button>
      <button>
        -
      </button>
    </div>
  </div>
</div>
`;
